A79-non-per-call-metrics-architecture.md

```diff
@@ -0,0 +1,598 @@
+ * MetricsRecorder abstracts away the complexity of handling multiple metrics 
+ * plugin from individual gRPC components.
+ */
+@Internal
+public interface MetricsRecorder {
+
+  /**
+   * Records a value for a long counter metric.
+   * 
+   * @param counterDescriptor The descriptor of the counter metric.
+   * @param value The value to record.
+   * @param labelValues  Required labels for identifying the metric.
+   * @param optionalLabelValues Additional labels to provide more context.
+   */
+  default void recordLongCounter(MetricDescriptor counterDescriptor, Long
+      value, List<String> labelValues, List<String> optionalLabelValues) {}
+
+  /**
+   * Records a value for a double histogram metric.
+   *
+   * @param histogramDescriptor The descriptor of the histogram metric.
+   * @param value The value to record.
+   * @param labelValues  Required labels for identifying the metric.
+   * @param optionalLabelValues Additional labels to provide more context.
+   */
+  default void recordDoubleHistogram(MetricDescriptor histogramDescriptor, 
+      Double value, List<String> labelValues, List<String> optionalLabelValues) 
+  {}
+}
+
+```
+
+##### Go
+
+(To be filled)
+
+### Metric Instrument Naming Conventions
+
+gRPC metrics follow the same general
+[guidelines](https://opentelemetry.io/docs/specs/semconv/general/metrics/#general-guidelines)
+as OpenTelemetry. The following additional guidelines apply -
+
+*   Metrics defined by gRPC should go under the `grpc.` namespace.
+*   Third-party plugins to gRPC should NOT use the `grpc.` namespace.
+*   Newer metrics should follow the same style as existing metrics.
+*   Metrics from load-balancing policies should be nested under `grpc.lb.`, e.g.
+    `grpc.lb.pick_first.`.
+*   Metrics from resolvers should be nested under `grpc.resolver.`, e.g.
+    `grpc.resolver.dns.`.
+*   Metrics from transports should be nested under `grpc.transport.`, e.g.,
+    `grpc.transport.http2.`
+
+### Metric Stability Guidelines and Guarantees
+
+From a stability perspective, metrics exported by gRPC will be in one of the
+following categories -
+
+*   Experimental, off by default
+*   Stable, off by default
+*   Stable, on by default
+
+Experimental metrics should always be off-by-default. gRPC reserves the right to
+change or remove these metrics. Documentation for the given metric determines
+whether the metric is experimental or non-experimental. An alternative approach
+of using an `experimental` namespace was considered, but not preferred since
+de-experimentalization of the metric would result in disruption of the
+corresponding metrics data (timeseries).
+
+New metrics should start as experimental. Once they are stable, they can
+optionally be made on-by-default. The distinction between on-by-default and
+off-by-default metrics allows gRPC to support metrics that may be desirable in
+some cases but too expensive to collect by default.
```

Comment:
I assume "too expensive to collect" is about CPU cost. But there is a storage cost argument as well, after it is . Are we only concerned with CPU cost here? (There could be memory costs, but in general I assume we won't have metrics that are too expensive in memory.)

Reply:
yeah, we are mostly worried about collection costs here which I assume would be mostly CPU, maybe memory as well, if, for example, the metric now requires to store some information for each RPC.